compiler/rustc_parse/src/lexer/mod.rs

```diff
@@ -322,6 +322,8 @@ impl<'a> StringReader<'a> {
                 let token = unicode_chars::check_for_substitution(self, start, c, &mut err);
                 if c == '\x00' {
                     err.help("source files must contain UTF-8 encoded text, unexpected null bytes might occur when a different encoding is used");
+                } else if let Ok(v) = &err.suggestions && v.len() == 0 && !c.is_ascii() {
```

Comment:
I think this needs a different heuristic for "is not visible when rendered", since this triggers on every non-ascii char.

Also, why are you checking `err.suggestions` here?

Reply:
> I think this needs a different heuristic for "is not visible when rendered", since this triggers on every non-ascii char.
> 
> Also, why are you checking `err.suggestions` here?

Yes, I'm still thinking any better method for check a unicode...

I'm checking the `err.suggestions` here, because we already have a strategy to give suggestions for those possible substitution, https://github.com/rust-lang/rust/blob/d76952fcbc418157b68dcdb42626b17f2348b8a9/compiler/rustc_parse/src/lexer/mod.rs#L322. 

My idea is if we already have such kind of suggestion, we add more, since it's clear enough:

```
 --> /home/cat/code/rust/src/test/ui/parser/unicode-quote-chars.rs:2:14
  |
2 |     println!(“hello world”);
  |              ^
  |
= help: Unicode character \u{323} might not be visible when rendered
help: Unicode characters '“' (Left Double Quotation Mark) and '”' (Right Double Quotation Mark) look like '"' (Quotation Mark), but are not
```

Reply:
> I'm checking the err.suggestions here, because we already have a strategy to give suggestions for those possible substitution, 

I think if you change your PR to only note characters that have a rendered width of zero (i.e. combining characters, etc) then this check is redundant.

Reply:
I haven't found a better way to figure out "not visible when rendered" chars.
We have a test case in https://github.com/rust-lang/rust/blob/b998821e4c51c44a9ebee395c91323c374236bbb/src/test/ui/issues/issue-29227.rs#L125

The `unicode_space_separator`, `unicode_space_separator` and `combining_spacing_mark` categories seem contain the chars which are hard to spot, but those are also not totaly 'not visible`, and I think introduce those big tables into compiler code base maybe too heavy ..